### PR TITLE
SNOW-200846 Fix NPE when multiple threads send OOB telemetry message.

### PR DESCRIPTION
**Pull request by sfc-gh-mrui**

### CODE REVIEW

**Review by sfc-gh-ema — APPROVED** [review]
Looks fine, have you checked that telemetry still works properly?

**Comment by sfc-gh-mrui** [comment]
> Looks fine, have you checked that telemetry still works properly?

@sfc-gh-ema I have confirmed the telemetry messages can be seen in the snowhouse.
@sfc-gh-sshankar Could you please merge this PR?

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala [hunk 1]
```diff
@@ -25,12 +25,20 @@ object SnowflakeTelemetry {
 
   private[snowflake] var output: ObjectNode = _
 
+  // The thread local TelemetryService need to be used because
+  // TelemetryService.getInstance returns a thread local instance.
+  private val threadLocalOobTelemetryService = new ThreadLocal[TelemetryService]() {
+    override protected def initialValue: TelemetryService = {
+      TelemetryService.enable()
```
[the remark this answers is not in the document]
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:32]: Java ThreadLocal is java feature which can guarantee it is initialized at most once for one thread.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala [hunk 2]
```diff
@@ -25,12 +25,20 @@ object SnowflakeTelemetry {
...
+      val service = TelemetryService.getInstance
```
[the remark this answers is not in the document]
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:33]: This is its implementation. JDBC TelemetryService uses ThreadLocal for the instance. Below is the code snippet:

public class TelemetryService {
  private static ThreadLocal<TelemetryService> _threadLocal =
      new ThreadLocal<TelemetryService>() {
        @Override
        protected TelemetryService initialValue() {
          return new TelemetryService();
        }
      };

  /** @return return thread local instance */
  public static TelemetryService getInstance() {
    return _threadLocal.get();
  }
}

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala [hunk 3]
```diff
@@ -100,7 +108,7 @@ object SnowflakeTelemetry {
       metric.put(TelemetryOOBFields.EXCEPTION_STACKTRACE, "NA")
     }
 
-    oobTelemetryService
+    val oobTelemetryService = getOobTelemetryService()
```
[the remark this answers is not in the document]
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:111]: It's thread safe. For different thread, getOobTelemetryService() will return different instance.

PS: ThreadLocal is new to me too. I have done some manual testing to make sure.
1. threadLocalOobTelemetryService.initialValue() is only called once, even one thread tries to send multiple messages.
2. Print the reference id of TelemetryService instance and the thread id to make sure different thread to access different TelemetryService instance.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (sfc-gh-sshankar) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:32]: No I mean is it ok for TelemetryService.enable() to be called multiple times.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfc-gh-sshankar) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:111]: I know about threadlocals, I am saying that I am not sure this is the right approach here. I think the correct thing to do is to make the report below thread safe.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:32]: I think it is safe to call TelemetryService.enable() multiple time. It is a synchronized function and only set a class static variable. 

  private static boolean enabled = true;
  private static final Object enableLock = new Object();
  public static void enable() {
    synchronized (enableLock) {
      enabled = true;
    }
  }

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:116]: Sorry for the misunderstanding. I will make the report() to be synchronized.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala [hunk 4]
```diff
@@ -100,7 +103,13 @@ object SnowflakeTelemetry {
       metric.put(TelemetryOOBFields.EXCEPTION_STACKTRACE, "NA")
     }
 
-    oobTelemetryService
+    // The constructor of TelemetryEvent.LogBuilder uses
+    // TelemetryService.getInstance to retrieve the thread safe instance
+    // and use it. So getOobTelemetryService() needs to be called before
+    // creating the LogBuilder object to make sure TelemetryService has
+    // been setup.
+    TelemetryService.enable()
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:111]: Why do you want to call this repeatedly ? Move this out into the initialization of the class.

**Review by sfc-gh-sshankar — APPROVED** [review]

**Pull request merged**